src/coreclr/jit/codegencommon.cpp

```diff
@@ -5869,6 +5870,7 @@ void CodeGen::genFnProlog()
             initReg  = genRegNumFromMask(tempMask);
         }
     }
+#endif
```

Comment:
For LoongArch64, it’s ok to keep `initReg = REG_SCRATCH` here.

Reply:
I do not see why this code should not run on LA64. Sounds like LA64 is not marking the reserved registers correctly.

Reply:
Hi, @jakobbotsch 
I try to explain it.
The `tempMash` is only used here to update the `initReg`.

I think these lines are old codes which only valid for x86/amd64.

Maybe it's best to refactor these codes based on the ISA is RISC or CISC.

Reply:
for arm64, the initial value of 'initReg' is REG_SCRATCH (in fact is REG_R9), while the `REG_PINVOKE_FRAME` is also REG_R9.
This is the root case why arm64 has to be updated liking the x86/amd64 when ` if (isRoot && compiler->opts.IsOSR())`.

for LA64, the `REG_SCRATCH` and `REG_PINVOKE_FRAME` are different.

Reply:
For RISC architectures liking LoongArch64, ARM64, RISCV64,   it's easy to avoid the conflict.

Reply:
@jakobbotsch 
which test case can test the `REG_PINVOKE_FRAME` ?

Reply:
Does the code cause any problems on LA64? It logically looks fine to me -- it is picking a register that it knows will be trashed so it can be used for initing.

Does it have to do with OSR? I can see that ARM64 and x64 have special handling for OSR below (that overrides `initReg` again) -- maybe add an LA64 case there as well? I would prefer it over fixing the same problem in a different way for LA64.

I agree it's likely that ARM64 does not _need_ this code, but `excludeMask` and `regSet.rsMaskResvd` can have registers in them for several reasons, and I am not confident those cannot end up with `REG_SCRATCH` in them in some cases where we then need to compensate for this.

Reply:
> Does the code cause any problems on LA64? It logically looks fine to me -- it is picking a register that it knows will be trashed so it can be used for initing.

The test cases of CoreRoot is OK for LA64.
But I just want to know the codes
https://github.com/dotnet/runtime/blob/613e4fecde90a05d1df0cda69392434514ac18f5/src/coreclr/jit/morph.cpp#L2734-L2738
how to return the `nonStdRegNum` as `REG_PINVOKE_FRAME`  ??
I tested the test caseds of CoreRoot and the library‘s tests, there is no this case on LA64.

 
> Does it have to do with OSR? 

Yes, the OSR for LA64 is tested OK.

Reply:
> Does the code cause any problems on LA64? It logically looks fine to me -- it is picking a register that it knows will be trashed so it can be used for initing.
> 
> Does it have to do with OSR? I can see that ARM64 and x64 have special handling for OSR below (that overrides `initReg` again) -- maybe add an LA64 case there as well? I would prefer it over fixing the same problem in a different way for LA64.
> 
> I agree it's likely that ARM64 does not _need_ this code, but `excludeMask` and `regSet.rsMaskResvd` can have registers in them for several reasons, and I am not confident those cannot end up with `REG_SCRATCH` in them in some cases where we then need to compensate for this.

Yes, these codes will cause the register conflict for large offset addresses of OSR root frames when `DOTNET_JitStressRegs=2` in `more_tailcalls.sh`.

The `initReg` was modified by `line:5870` from `REG_SCRATCH` to `REG_A0` on LA64. (ARM64 also has the same action, but ARM64 instruction `ldr` can handle bigger offset, and ARM64 has special handling for OSR below, so ARM64 didn’t meet the problem.)

Currently we didn’t meet any problem test cases with this solution (including CoreRoot testcases、CoreRoot stress tests、libs.tests、aspnetcore testcases). This solution can get a little bit gain for LA64.

Reply:
`REG_PINVOKE_FRAME` looks unused on 64-bit platforms. That doesn't mean that LSRA cannot assign `REG_SCRATCH` to something else, however (at least on arm64. I don't know about LA64.)

> ARM64 has special handling for OSR below, so ARM64 didn’t meet the problem

Please switch LA64 to be similar to x64 and ARM64. x64 and ARM64 both override `initReg` below for OSR methods.

Reply:
> I am suggesting to do the following:
> 
> 1. Remove `#ifndef TARGET_LOONGARCH64` here

By the way, I think these old codes about the conflics 
https://github.com/dotnet/runtime/blob/613e4fecde90a05d1df0cda69392434514ac18f5/src/coreclr/jit/codegencommon.cpp#L5891-L5895

might be cased here
https://github.com/dotnet/runtime/blob/613e4fecde90a05d1df0cda69392434514ac18f5/src/coreclr/jit/codegencommon.cpp#L6234-L6242

But these is not used by LA64 and the LA64 can hold the large offset.


> 2. Add `#ifdef TARGET_LOONGARCH64` below and handle OSR methods like arm64 and x64 is handled (`if (isRoot && compiler->opts.IsOSR()) { initReg = REG_SCRATCH; }`)
> 
> The point is to align the platforms with the same behavior. It will make it easier in the future if we decide to clean this up a bit.

OK. We will update it.

Reply:
I agree that the assert looks outdated. Feel free to update it to `defined(TARGET_X86) || defined(TARGET_ARM)`.

Reply:
Already updated, @shushanhf  @jakobbotsch  thank you!